source/fuzz/transformation_wrap_vector_synonym.h

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2021 Shiyu Liu
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_FUZZ_TRANSFORMATION_WRAP_VECTOR_SYNONYM_H_
+#define SOURCE_FUZZ_TRANSFORMATION_WRAP_VECTOR_SYNONYM_H_
+
+#include "source/fuzz/protobufs/spirvfuzz_protobufs.h"
+#include "source/fuzz/transformation.h"
+#include "source/fuzz/transformation_context.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace fuzz {
+
+class TransformationWrapVectorSynonym : public Transformation {
+ public:
+  explicit TransformationWrapVectorSynonym(
+      protobufs::TransformationWrapVectorSynonym message);
+
+  TransformationWrapVectorSynonym(uint32_t instruction_id,
+                                  uint32_t vector_operand1,
+                                  uint32_t vector_operand2, uint32_t fresh_id,
+                                  uint32_t pos);
+  // - |instruction_id| must be the id of a arithmetic operation.
+  // - |vector_operand1| and |vector_operand2| represents the result ids of the
+  //   two added vector.
+  // - |fresh_id| is a vector type for the result of the transformation.
+  // - result vector type must match the type of two vectors being added.
+  // - |fresh_id| must be fresh.
+  // - |vector_operand1| and |vector_operand2| must have the same vector type.
+  // - |pos| is a 0-indexed position of the component that contains the
+  //   value of a and b. pos must be smaller than the number of
+  //   elements that the vector type can has.
+  bool IsApplicable(
+      opt::IRContext* ir_context,
+      const TransformationContext& transformation_context) const override;
+
+  void Apply(opt::IRContext* ir_context,
+             TransformationContext* transformation_context) const override;
+
+  std::unordered_set<uint32_t> GetFreshIds() const override;
+  protobufs::Transformation ToMessage() const override;
+
+  static bool OperandTypeIsSupported(opt::Instruction* type_instruction) {
+    return (type_instruction->opcode() == SpvOpTypeFloat ||
+            type_instruction->opcode() == SpvOpTypeInt) &&
+           type_instruction->GetSingleWordInOperand(0) == 32;
```

Comment:
Why do we need to restrict only to 32-bit numbers? @afd.

Reply:
I had thought that SPIR-V only allowed vectors over 32-bit types.

But looking at the spec I cannot find any mention of this.

@sliu-UIUC Could you please write a SPIR-V test case that uses 64-bit integer and floating-point types, and check that vector wrapping works as expected when 64-bit types are used? I believe you'll need to use some capabilities to enable 64-bit scalar types.

If all works as it should then indeed the requirement of 32 bits here can go. The new test case should be included in the test file.

Reply:
Done.